### DISCUSSION TITLE
App::wait() and coherence within multiple event() invocations

### DISCUSSION THREAD

**Discussion by saflo76** [fltk-rs/fltk-rs#1600] [Q&A]
Hi All,
In my project after the invocation of App::wait() inside the event loop, I tend to have the need to call functions like event() and event_coords() multiple times inside methods (usually it happens within program sub-calls), this during the single event loop lapse (between a wait() call and the next one).
Since usually in my implementation a sequence of nested method calls work like a progressive process to reach the real meant routine to process the event (for ex. mouse click inside specific sub-area), the need of a coherence for keyboard/mouse probed status among multiple event probing calls (event(), event_coords(), etc...) invocations in important, otherwise I cannot analyze a precise event condition.
This is respected if wait() method after probing input data, keeps the result while stacking the next events in a buffer to release them only in each next wait() invocation.
Otherwise to keep a coherent input scenario I need to save keyboard/mouse event/status in a shared state struct and work quite a bit around it to implement it as effective as possible, as many spreaded methods inside my program will need to access the SAME input event and state.

How does really behave this under the hood in FLTK?
Thanks
Sandro

**Comment by MoAlyousef** [comment]
Hello

app::wait ultimately calls Fl::wait which waits for application and systems events, timers, polls file descriptors and sockets in some systems etc. FLTK also processes there the event queue and flushes the graphics context. 
That said, it is not where the events (which can be queried using event() and event_* functions) are set. They are set for each window in the window's handler, then the events are dispatched to the window's children if not handled directly by the window.
So as you have deduced, when you call event() for example inside app::wait(), the value you get will be the last event arriving to the event loop, which might not be what you're waiting for. 
To get live events you would need to override the window handler itself. The following program shows the difference between handling things in the event loop vs in the window handler:
```rust
use fltk::{prelude::*, enums::*, *};

fn main() {
    let a = app::App::default();
    let mut w = window::Window::default().with_size(400, 300);
    w.end();
    w.show();
    w.handle(|w, ev| {
        println!("window event: {:?}", ev);
        false
    });
    while a.wait() {
        println!("event loop: {:?}", app::event());
    }
}
```

**Reply by saflo76** [reply]
Thanks for your reply.
So an event() call within main loop can only catch those events previously reported as unmanaged (handler(closure -> false)) by various handlers, which normally process them for first. If I have understood correctly in your example inside the handle() closure, by reporting always false any event remains "suspended" so 100% of them can be pulled by event() call in main loop.

This is good to know, however my problem is knowing if when calling event() and any event_*() more times closely I may get different results or a same pending one until another call of wait() is done (sort of like wait() contemplating 1 event per call atomically).

I ask this because for ex. I need to examine mouse event and position in many routines, one subsequently calling the other, but by working on the same initial poll values, so I'm starting to guess that I need to save the state of input from one initial event() / event_*() call and avoid any further invocation until cycle is done.
Of course managing state in a shared but exclusive way (Rc<RefCell<>>) accessible by the various involved routines.

**Reply by MoAlyousef** [reply]
The event() inside the event loop is not necessarily unhandled. It might just be a different snapshot. The top window handler is what sets these events in the first place, so if you need live events, the handler is where you would want to listen for them.
For example, in the fltk-egui crate, the [event propagation code](https://github.com/fltk-rs/fltk-egui/blob/f20a2e3294dede4612356959f514000181b75f84/src/lib.rs#L169) is called from the window handler:
https://github.com/fltk-rs/fltk-egui/blob/f20a2e3294dede4612356959f514000181b75f84/examples/basic.rs#L24

There's a simple example in the fltk-rs repo, a pong game which also shows such usage:
https://github.com/fltk-rs/fltk-rs/blob/master/fltk/examples/pong.rs

The idea is that if you need to store the event (for example in an Rc RefCell as you suggested), it might be overwritten by another such as is happening when you query an event in the event loop. 

Now you don't need to always override the top window handler, the widget handler might be sufficient if you only need such realtime handling for said widget.
